src/mame/yamaha/ymdx100.cpp

```diff
@@ -92,20 +306,264 @@ void yamaha_dx100_state::mem_map(address_map &map)
+	// simulate not having a breath controller attached at all???
+	PORT_BIT( 0xff, 0, IPT_PEDAL ) PORT_NAME("Breath Controller") PORT_SENSITIVITY(100) PORT_KEYDELTA(10) PORT_MINMAX(0x00, 0xff) PORT_REVERSE
+
+	PORT_START("AN3")
+	// The data entry slider stays in place to wherever it's set.
+	PORT_BIT( 0xff, 0, IPT_POSITIONAL_V ) PORT_NAME("Data Entry Slider") PORT_SENSITIVITY(100) PORT_KEYDELTA(10) PORT_MINMAX(0x00, 0xff) PORT_CENTERDELTA(0)
+
+	PORT_START("AN5")
+	PORT_CONFNAME( 0xff, 0x80, "Internal RAM Battery Level" )
+	// "CNG RAM BATTERY!" displayed unless value is between 0x70 and 0xcc
+	PORT_CONFSETTING( 0x6f, "Too Low" )
+	PORT_CONFSETTING( 0x70, "Lowest Allowed" )
+	PORT_CONFSETTING( 0x80, "Normal" ) // for some arbitrary definition of "normal"
+	PORT_CONFSETTING( 0xcb, "Highest Allowed" )
+	PORT_CONFSETTING( 0xcc, "Too High" )
+
+	PORT_START("AN6")
+	PORT_CONFNAME( 0xff, 0x00, "Battery Power Level" )
+	// Yes, higher values mean lower voltages.
+	// I think this is opposite to how the RAM battery voltmeter works.
+	// The weird granularity here is due to the buggy implementation of the
+	// test in the test mode; all 7V values should LED flash, but some don't.
+	PORT_CONFSETTING( 0x00, "9V (Normal)" ) // for some arbitrary definition of "normal"
+	PORT_CONFSETTING( 0x4b, "9V (Lowest Allowed)" )
+	PORT_CONFSETTING( 0x67, "7V (Highest Allowed Without LED Flash)" )
+	PORT_CONFSETTING( 0x6b, "7V (Lowest Allowed Without LED Flash)" )
+	PORT_CONFSETTING( 0x6c, "7V (Highest Allowed With LED Flash)" )
+	PORT_CONFSETTING( 0x6f, "7V (Lowest Allowed With LED Flash)" )
+	PORT_CONFSETTING( 0x70, "Less Than 7V" )
 INPUT_PORTS_END
 
 void yamaha_dx100_state::dx100(machine_config &config)
 {
 	HD6303X(config, m_maincpu, 7.15909_MHz_XTAL / 2); // HD6303XP
 	m_maincpu->set_addrmap(AS_PROGRAM, &yamaha_dx100_state::mem_map);
-	m_maincpu->in_p2_cb().set([this]() -> u8 { return m_midi_in ? 0x04 : 0; });
-	m_maincpu->in_p6_cb().set("adc", FUNC(m58990_device::eoc_r)).lshift(4);
+	m_maincpu->in_p2_cb().set_ioport("P2");
+	m_maincpu->out_p2_cb().set_ioport("P2");
+	m_maincpu->in_p6_cb().set([this]() -> u8 {
+		// [TODO1] The threshold here and the output values below should be double-checked.
+		u8 casplay = (m_cassette->input() > 0.009) ? 0x80 : 0;
+		// [TODO1] Need to verify if this bit is held on output or not
+		u8 casrec = m_port6_val & 0x40;
+		u8 adcval = ((u8) (m_adc->eoc_r() << 4)) & 0x10;
+		u8 port5line = (m_port6_val & 0x2f);
+		return casplay | casrec | adcval | port5line;
+	});
+	m_maincpu->out_p6_cb().set([this](u8 val) {
+		m_port6_val = val;
+		m_cassette->output(((m_port6_val & 0x40) != 0) ? -1.0 : 1.0);
+	});
+	m_maincpu->in_p5_cb().set([this]() -> u8 {
+		u8 line = m_port6_val & 0x2f;
+		if ((line & 0x20) != 0)
+			return 0x00;	// [TODO2]
+		return (u8) (m_keysbuttons[line]->read() & 0xff);
+	});
 	m_maincpu->out_ser_tx_cb().set("mdout", FUNC(midi_port_device::write_txd));
 
+	m_port6_val = 0;		// TODO figure out the actual power-on state
```

Comment:
This should go to `machine_reset` fn, definitely not in machine_config

Reply:
Done now (will push later after resolving the rest)